jdbc/src/main/java/com/interface21/jdbc/datasource/DataSourceUtils.java

```diff
@@ -1,33 +1,43 @@
 package com.interface21.jdbc.datasource;
 
-import com.interface21.jdbc.CannotGetJdbcConnectionException;
-import com.interface21.transaction.support.TransactionSynchronizationManager;
-
-import javax.sql.DataSource;
 import java.sql.Connection;
 import java.sql.SQLException;
 
-// 4단계 미션에서 사용할 것
+import javax.sql.DataSource;
+
+import com.interface21.jdbc.CannotGetJdbcConnectionException;
+import com.interface21.transaction.support.TransactionSynchronizationManager;
+
 public abstract class DataSourceUtils {
 
-    private DataSourceUtils() {}
+    private DataSourceUtils() {
+    }
 
-    public static Connection getConnection(DataSource dataSource) throws CannotGetJdbcConnectionException {
+    public static Connection getConnection(final DataSource dataSource) throws CannotGetJdbcConnectionException {
         Connection connection = TransactionSynchronizationManager.getResource(dataSource);
         if (connection != null) {
             return connection;
         }
 
         try {
-            connection = dataSource.getConnection();
-            TransactionSynchronizationManager.bindResource(dataSource, connection);
-            return connection;
+            return dataSource.getConnection();
         } catch (SQLException ex) {
             throw new CannotGetJdbcConnectionException("Failed to obtain JDBC Connection", ex);
         }
     }
 
-    public static void releaseConnection(Connection connection, DataSource dataSource) {
+    public static void releaseConnection(final Connection connection, final DataSource dataSource) {
+        if (connection == null) {
+            return;
+        }
+
+        Connection txSyncConnection = TransactionSynchronizationManager.getResource(dataSource);
+        if (txSyncConnection != null && txSyncConnection == connection) {
+            // 트랜잭션 커넥션인 경우 닫지 않음 (트랜잭션 매니저가 관리)
+            return;
+        }
```

Comment:
저랑 비슷하게 해결하셨네요~

트랜잭션 커넥션 검사를 별도 메서드로 분리할 수 있을 것 같아요. 혹은 아래와 같이도 가능할 것 같습니다!


```suggestion
        Connection isXX = TransactionSynchronizationManager.hasResource(dataSource);
        if (isXX) {
            return;
        }
```

Reply:
메서드 분리로 읽기 쉬운 코드로 작성할 수 있네요. 👍